# **B KNN MNIST**

What is the error rate of KNN on the test set?

What is the error rate for each label (number)?

Do for k = 2, 4, 8

How does the choice of k influence the results?

In [0]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.manifold import TSNE
import scipy.spatial.distance
import scipy
import pandas as pd
from sklearn.metrics import f1_score


plt.set_cmap('gray')

<Figure size 432x288 with 0 Axes>

In [0]:
def data_preparation():
  #Trainings- und Testdaten
  mnist_train = tf.keras.datasets.mnist.load_data()[0]
  digits_train, labels_train = mnist_train[0], mnist_train[1]

  mnist_test = tf.keras.datasets.mnist.load_data()[1]
  digits_test, labels_test = mnist_test[0], mnist_test[1]
  
  #Anzahl der Samples in train und test
  n_train = digits_train.shape[0]
  n_test = digits_test.shape[0]

  #Anzahl der Pixel
  m = digits_train.shape[1]**2

  #Vektor aus Matrizen erstellen -> digits flat
  digits_train = digits_train.reshape([n_train, m])
  digits_test = digits_test.reshape([n_test, m])

  #Subsamples der unserer Datensätze
  idx = np.random.randint(n_train, size = 60000)
  global digits_train_sample, labels_train_sample
  digits_train_sample = digits_train[idx]
  labels_train_sample = labels_train[idx]

  global digits_test_sample, labels_test_sample
  idx = np.random.randint(n_test, size = 10000)
  digits_test_sample = digits_test[idx]
  labels_test_sample = labels_test[idx]

In [0]:
def knn(k):
  #Entfernugen berechnen. Man erhält als Matrix alle Entfernungen und kann dann die k nächsten Nachbarn ermitteln
  #k = 2
  print("KNN für k=", k)
  #Distanzen 
  dists = scipy.spatial.distance.cdist(digits_train_sample, digits_test_sample, metric='euclid')
  #index der nächsten Nachbarn
  idx_nearest = np.argpartition(dists, k, axis=0)[:k]
  nearest_dists = np.take(digits_train_sample, idx_nearest)

  knn_neighbors = labels_train_sample[idx_nearest]
  #remove count array with [0]
  knn_result = scipy.stats.mode(knn_neighbors)[0]
  knn_result = knn_result[0]
  return knn_result

In [0]:
def calc_error(knn_result):
  n_error = 0
  check_matrix = labels_test_sample == knn_result
  for check_bool in check_matrix:
    if check_bool == False:
      n_error = n_error + 1
  error_rate = n_error / len(knn_result) * 100
  print("Errorrate: ", error_rate, "%")

  error = {0:0,1:0,2:0,3:0,4:0,5:0,6:0,7:0,8:0,9:0}
  total ={0:0,1:0,2:0,3:0,4:0,5:0,6:0,7:0,8:0,9:0}

  for i in range(len(labels_test_sample)):
    if labels_test_sample[i] != knn_result[i]:
      error[labels_test_sample[i]] +=1
      total[labels_test_sample[i]] +=1
    else:
      total[labels_test_sample[i]] +=1
  
  errorrate = {}
  for i in range(0,10):
    errorrate[i] = (error[i] / total[i]) * 100
    print("Errorrate für", i, " = ", round(errorrate[i], 2), "%")
  print(pd.DataFrame([error]))
  print(pd.DataFrame([total]))

In [0]:
#for testing
def calc_f1(knn_result):
  f1 = f1_score(labels_test_sample, knn_result, average="micro")
  print("Errorrate", (1 - f1))

In [0]:
data_preparation()
knn_result = knn(2)
calc_error(knn_result)

print("________________________")

knn_result = knn(4)
calc_error(knn_result)

print("________________________")

knn_result = knn(8)
calc_error(knn_result)


KNN für k= 2
Errorrate:  4.38 %
Errorrate für 0  =  1.13 %
Errorrate für 1  =  0.0 %
Errorrate für 2  =  5.41 %
Errorrate für 3  =  3.88 %
Errorrate für 4  =  3.45 %
Errorrate für 5  =  5.17 %
Errorrate für 6  =  2.06 %
Errorrate für 7  =  4.39 %
Errorrate für 8  =  11.32 %
Errorrate für 9  =  6.75 %
    0  1   2   3   4   5   6   7    8   9
0  11  0  58  38  32  47  19  46  114  73
     0     1     2    3    4    5    6     7     8     9
0  976  1078  1073  979  927  909  923  1047  1007  1081
________________________
KNN für k= 4
Errorrate:  4.37 %
Errorrate für 0  =  1.13 %
Errorrate für 1  =  0.0 %
Errorrate für 2  =  5.96 %
Errorrate für 3  =  4.19 %
Errorrate für 4  =  4.1 %
Errorrate für 5  =  4.51 %
Errorrate für 6  =  1.95 %
Errorrate für 7  =  4.2 %
Errorrate für 8  =  10.72 %
Errorrate für 9  =  6.66 %
    0  1   2   3   4   5   6   7    8   9
0  11  0  64  41  38  41  18  44  108  72
     0     1     2    3    4    5    6     7     8     9
0  976  1078  1073  979  927  909 

# **Antwort**

Basierend auf 10000 Testsamples, da keine Größe angegeben war ("on the test set").

**What is the error rate of KNN on the test set and what is the error rate for each label (number)?**

Errorrate = Falsche Predictions / Testsamples

KNN für k= 2

```
Errorrate auf Test Set:  4.38 %
    
Errorrate für 0  =  1.13 %
Errorrate für 1  =  0.0 %
Errorrate für 2  =  5.41 %
Errorrate für 3  =  3.88 %
Errorrate für 4  =  3.45 %
Errorrate für 5  =  5.17 %
Errorrate für 6  =  2.06 %
Errorrate für 7  =  4.39 %
Errorrate für 8  =  11.32 %
Errorrate für 9  =  6.75 %
```
----

KNN für k= 4
```
Errorrate auf Test Set:  4.37 %

Errorrate für 0  =  1.13 %
Errorrate für 1  =  0.0 %
Errorrate für 2  =  5.96 %
Errorrate für 3  =  4.19 %
Errorrate für 4  =  4.1 %
Errorrate für 5  =  4.51 %
Errorrate für 6  =  1.95 %
Errorrate für 7  =  4.2 %
Errorrate für 8  =  10.72 %
Errorrate für 9  =  6.66 %
```
----

KNN für k= 8
```
Errorrate auf Test Set:  4.17 %
    
Errorrate für 0  =  1.43 %
Errorrate für 1  =  0.28 %
Errorrate für 2  =  6.24 %
Errorrate für 3  =  3.98 %
Errorrate für 4  =  3.88 %
Errorrate für 5  =  4.51 %
Errorrate für 6  =  2.17 %
Errorrate für 7  =  4.01 %
Errorrate für 8  =  9.43 %
Errorrate für 9  =  5.55 %
```
**How does the choice of k influence the results?** 

Die Errorrate auf dem Test Set ändert sich nur minimal und nimmt von k = 2 auf k = 4 um 0.01% ab und von k = 4 auf k = 8 um 0.2% ab. Dies ist angesichts der größe des Test Sets eine bemerkbare Veränderung, jedoch haben sich die Ergebnisse für die einzelnen Zahlen teils deutlich mehr verändert. 
Für die Zahl 0 steigt die Errorrate von k=2 1.13% auf k=8 1.43%.
Für die Zahl 8 sinkt die Errorrate von k=2 11.32% auf k=8 9.43 %.
Wie man anhand der Errorrate erkennen kann ist die Tedenz, dass mit k=8 genauere Ergebnisse erzielt werden können. Wenn man jedoch die einzelnen Zahlen betrachtet, haben manche Zahlen wie z.B. die Zahl 1 mit k=8 eine höhere Errorrate als mit k=2.  
Diese Deltas scheinen klein zu sein, jedoch machen sie bei großen Datensätzen den Unterschied zwischen einem gutem oder schlechtem Algorithmus aus. 

Ebenfalls ist anzumerken: Wenn nicht 10000 Testsamples benutzt werden sind die Schwankungen nochmal deutlich größer.